KSGraphVertex.java

```diff
@@ -0,0 +1,269 @@
+        }
+        else
+        {
+            // we should throw an exception
+
+        }
+    }
+
+    public ArrayList<KSGraphVertex> ShortestPathUnWeighted(KSGraphVertex targetVertex)
+    {
+        Queue<KSGraphVertexWeighted> vertexQueue = new LinkedList<KSGraphVertexWeighted>();
+        HashMap<String, KSGraphVertexBackTrack> queuedVertexes = new HashMap<String, KSGraphVertexBackTrack>();
+        //ArrayList<ArrayList<KSGraphVertex>> paths = new ArrayList<ArrayList<KSGraphVertex>>();
+        //ArrayList<KSGraphVertex> path = new ArrayList<KSGraphVertex>();
+        KSGraphVertexWeighted clonedGraphVertexWeighted = new KSGraphVertexWeighted(this);
+        queuedVertexes.put(clonedGraphVertexWeighted.GetVertexName(), new KSGraphVertexBackTrack(clonedGraphVertexWeighted));
+
+        vertexQueue.add(clonedGraphVertexWeighted);
+
+        while(!vertexQueue.isEmpty())
+        {
+            KSGraphVertexWeighted currentVertex = vertexQueue.remove();
+            KSGraphVertexWeighted newGraphVertex = null;
+            KSGraphVertexBackTrack newGraphVertexBackTrack = null;
+
+            for(KSGraphEdge edge : currentVertex.GetOutgoingEdges())
+            {
+                if(queuedVertexes.containsKey(edge.GetTargetVertex().GetVertexName()))
+                {
+                    newGraphVertexBackTrack = queuedVertexes.get(edge.GetTargetVertex().GetVertexName());
+                    newGraphVertex = newGraphVertexBackTrack.GetTargetVertex();
+                    if(newGraphVertex.GetVertexWeight()>currentVertex.GetVertexWeight()+1)
+                    {
+                        newGraphVertex.SetVertexWeight(currentVertex.GetVertexWeight()+1);
+                        newGraphVertexBackTrack.GetBackTrackList().clear();
+                        newGraphVertexBackTrack.GetBackTrackList().addAll(queuedVertexes.get(currentVertex.GetVertexName()).GetBackTrackList());
+                        newGraphVertexBackTrack.GetBackTrackList().add(currentVertex);
+                    }
+                }
+                else
+                {
+                    newGraphVertex = (KSGraphVertexWeighted) edge.GetTargetVertex();
+                    KSGraphVertexBackTrack backTrack = new KSGraphVertexBackTrack(newGraphVertex);
+                    backTrack.GetBackTrackList().addAll(queuedVertexes.get(currentVertex.GetVertexName()).GetBackTrackList());
+                    backTrack.GetBackTrackList().add(currentVertex);
+                    queuedVertexes.put(newGraphVertex.GetVertexName(), backTrack);
+                    newGraphVertex.SetVertexWeight(currentVertex.GetVertexWeight()+1);
+                }
+
+                vertexQueue.add((KSGraphVertexWeighted) edge.GetTargetVertex());
+            }
+        }
+
+        //
+        // CodeGuru should have ensured that we are not de-referencing a null pointer
+        //
+        return queuedVertexes.get(targetVertex.GetVertexName()).GetBackTrackList();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;targetVertex&#39; (index: 0 | type: KSGraphVertex). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;ConnectToTarget&#39;. The same parameter type is validated here for example: KSGraphVertex.java:200. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).